### ISSUE TITLE
ci: test with miri

### ISSUE THREAD

**Issue by storopoli** [rust-bitcoin/rust-bitcoin#3071]
We really should. I tested this locally.

_Originally posted by @Kixunil in https://github.com/rust-bitcoin/rust-bitcoin/issues/3066#issuecomment-2233593601.

Should we test our CI with `miri`?

A good template from Jonhoo: https://github.com/jonhoo/rust-ci-conf/blob/799c99e0cf77d2289ab47e84fa3b8296cfc5beb1/.github/workflows/safety.yml#L55-L71

---

From [`rust-lang/miri`](https://github.com/rust-lang/miri)

Miri is an Undefined Behavior detection tool for Rust. It can run binaries and test
suites of cargo projects and detect unsafe code that fails to uphold its safety requirements. For
instance:

* Out-of-bounds memory accesses and use-after-free
* Invalid use of uninitialized data
* Violation of intrinsic preconditions (an [`unreachable_unchecked`] being
  reached, calling [`copy_nonoverlapping`] with overlapping ranges, ...)
* Not sufficiently aligned memory accesses and references
* Violation of basic type invariants (a `bool` that is not 0 or 1, for example,
  or an invalid enum discriminant)
* **Experimental**: Violations of the [Stacked Borrows] rules governing aliasing
  for reference types
* **Experimental**: Violations of the [Tree Borrows] aliasing rules, as an optional
  alternative to [Stacked Borrows]
* **Experimental**: Data races and emulation of weak memory effects, i.e.,
  atomic reads can return outdated values.

**Comment by Kixunil** [comment]
Relevant limitations for us:

* missing `sha` intrinsics until my PR is merged (but we could use my branch if we want to or merge #3066)
* whole `secp256k` with FFI. (maybe we could have full reimpl behind `#[cfg(miri)]`?)
* any tests involving files/TCP (I think I saw them; we could cfg-them out or rewrite)

However, what we mainly need is usages of `unsafe` - there are a few in `hashes`, in `internals` and in `script` (maybe a reason to make it a crate? or to make `keys` optional in `primitives`?) There's a ton in `secp256k1` but we can't fix or `miri`-test that. I also have ideas to use unsized types more which will probably involve small bits of `unsafe`.

**Comment by apoelstra** [comment]
Definitely I'd like to use miri everywhere we can. A long time ago I think I ran the full test suite through Miri, which was pretty slow, but it was a bunch of work to blacklist everything that called into libsecp functions. I don't think it occured to me to try to do this in CI.

**Comment by Kixunil** [comment]
Yeah, it's very slow. Just sha256 takes several minutes. We'll need to put it into cron and split up.

**Comment by jrakibi** [comment]
I think this can be closed. Miri was already added to CI, and now it runs as a cron job.

**Comment by jrakibi** [comment]
It might be worth also testing an AArch64 target at some point, tho Miri is already quite slow on the current x86

**Comment by Kixunil** [comment]
Thanks for heads up!

> AArch64 target at some point, tho Miri is already quite slow on the current x86

Notably, it'd be equally slow, not slower because Miri can itself emulate other architectures. Also does Miri support AArch64 SHA256 intrinsics already?

**Comment by jrakibi** [comment]
> does Miri support AArch64 SHA256 intrinsics already?

I opened a PR yesterday to add support for that: https://github.com/rust-lang/miri/pull/5064

Hopefully it gets merged, cause I hit UB when running the `bitcoin_hashes` tests under my local Miri branch that supports AArch64 SHA256 intrinsics: https://github.com/rust-bitcoin/rust-bitcoin/pull/6262
